ic-agent/src/agent/mod.rs

```diff
@@ -446,14 +459,80 @@ impl Agent {
         effective_canister_id: Principal,
         signed_query: Vec<u8>,
     ) -> Result<Vec<u8>, AgentError> {
-        let _envelope: Envelope =
+        let envelope: Envelope =
             serde_cbor::from_slice(&signed_query).map_err(AgentError::InvalidCborData)?;
-        self.query_endpoint::<QueryResponse>(effective_canister_id, signed_query)
-            .await
-            .and_then(|response| match response {
-                QueryResponse::Replied { reply } => Ok(reply.arg),
-                QueryResponse::Rejected(response) => Err(AgentError::ReplicaError(response)),
-            })
+        self.query_inner(
+            effective_canister_id,
+            signed_query,
+            envelope.content.to_request_id(),
+        )
+        .await
+    }
+
+    /// Helper function for performing both the query call and possibly a read_state to check the subnet node keys.
+    ///
+    /// This should be used instead of `query_endpoint`. No validation is performed on `signed_query`.
+    async fn query_inner(
+        &self,
+        effective_canister_id: Principal,
+        signed_query: Vec<u8>,
+        request_id: RequestId,
+    ) -> Result<Vec<u8>, AgentError> {
+        let response = if self.verify_query_signatures {
+            let (response, subnet) = futures_util::try_join!(
+                self.query_endpoint::<QueryResponse>(effective_canister_id, signed_query),
+                self.get_subnet_by_canister(&effective_canister_id)
+            )?;
+            if response.signatures().is_empty() {
+                return Err(AgentError::MissingSignature);
+            }
+            for signature in response.signatures() {
+                if OffsetDateTime::now_utc()
+                    - OffsetDateTime::from_unix_timestamp_nanos(signature.timestamp as _).unwrap()
```

Comment:
Can we return an error rather than panic here?

Reply:
A u64 nanosecond timestamp is always within range; this call cannot panic without a breaking change in `time`.